keras/src/quantizers/gptqutils.py

```diff
@@ -0,0 +1,386 @@
+        text_list = [d[text_column] for d in raw_dataset]
+        full_text = "\n\n".join(text_list)
+        all_tokens = tokenizer.tokenize(full_text)
+
+    else:
+        logging.info("\n==> Using pre-made dataset/generator...")
+        dataset_list = list(dataset)
+
+        if not dataset_list:
+            raise ValueError("Provided dataset is empty.")
+
+        if isinstance(dataset_list[0], str):
+            logging.info("   (Dataset contains strings, tokenizing now...)")
+            full_text = "\n\n".join(dataset_list)
+            all_tokens = tokenizer.tokenize(full_text)
+        else:
+            logging.info("   (Dataset is pre-tokenized, concatenating...)")
+            concatenated_tokens = ops.concatenate(
+                [ops.reshape(s, [-1]) for s in dataset_list], axis=0
+            )
+            all_tokens = ops.convert_to_numpy(concatenated_tokens)
+
+    all_tokens = np.array(all_tokens, dtype=np.int32)
+
+    # --- Step 2: Repeat data if it's too short ---
+    required_tokens = nsamples * seqlen
+    if len(all_tokens) < required_tokens:
+        logging.info(
+            f"Warning: Dataset is too short ({len(all_tokens)} tokens)."
+            " Repeating data to generate {nsamples} samples."
+        )
+        repeats = -(-required_tokens // len(all_tokens))  # Ceiling division
+        all_tokens = np.tile(all_tokens, repeats)
+
+    # --- Step 3: Chunk the token list into samples ---
+    # utils.set_random_seed(seed)
+
+    calibration_samples = []
+    for _ in range(nsamples):
+        # Generate a random starting index
+        start_index = random.randint(0, len(all_tokens) - seqlen - 1)
+        end_index = start_index + seqlen
+        sample = all_tokens[start_index:end_index]
+        calibration_samples.append(ops.reshape(sample, (1, seqlen)))
+
+    final_array = ops.stack(calibration_samples, axis=0)
+    return ops.convert_to_numpy(final_array)
+
+
+def _find_layers_recursive(layer, prefix, found_layers):
+    """
+    Recursively search for Dense and EinsumDense layers and record them.
+    """
+    for sub_layer in layer._layers:
+        # Construct a unique name for the layer based on its hierarchy
+        layer_name = f"{prefix}.{sub_layer.name}"
+        if isinstance(sub_layer, (Dense, EinsumDense)):
+            found_layers[layer_name] = sub_layer
+
+        # Recurse into nested layers that are not the target types
+        elif hasattr(sub_layer, "_layers") and sub_layer._layers:
+            _find_layers_recursive(sub_layer, layer_name, found_layers)
+
+
+def find_layers_in_block(block):
+    """
+    A pluggable, generic function to find all Dense and EinsumDense layers
+    within any transformer block by using a recursive search.
+    """
+    found_layers = {}
+    # Start the recursive search from the block itself
+    _find_layers_recursive(block, "block", found_layers)
+    return found_layers
+
+
+def apply_gptq_layerwise(
```

Comment:
Let's add a docstring here. It might also be a good idea to break this out into smaller methods, the existing implementation is a little too long for a single function.